File: BringYourOwn-Samples/Src/VideoIndexerClient/auth/AccountTokenProvider.cs
using System.Net.Http.Headers;
using System.Text.Json;
using Azure.Core;
using Azure.Identity;
using Microsoft.Extensions.Logging;
using static VideoIndexerClient.Utils.Consts;

namespace VideoIndexerClient.auth
{

    public static class AccountTokenProvider
    {
        public static async Task<string> GetAccountAccessTokenAsync(ILogger logger, ArmAccessTokenPermission permission = ArmAccessTokenPermission.Contributor, ArmAccessTokenScope scope = ArmAccessTokenScope.Account, CancellationToken ct = default)
        {
            var armAccessToken = await GetArmAccessTokenAsync(ct);
            var accountAccessToken = await GetAccountAccessTokenAsync(logger,armAccessToken, permission, scope, ct);
            return accountAccessToken;
        }

        public static async Task<string> GetArmAccessTokenAsync(CancellationToken ct = default)
        {
            var tokenRequestContext = new TokenRequestContext(new[] { $"{AzureResourceManager}/.default" });
            var tokenRequestResult = await new DefaultAzureCredential().GetTokenAsync(tokenRequestContext, ct);
            return tokenRequestResult.Token;
        }

        public static async Task<string> GetAccountAccessTokenAsync(ILogger logger, string armAccessToken, ArmAccessTokenPermission permission = ArmAccessTokenPermission.Contributor, ArmAccessTokenScope scope = ArmAccessTokenScope.Account, CancellationToken ct = default)
        {
            var accessTokenRequest = new AccessTokenRequest
            {
                PermissionType = permission,
                Scope = scope
            };

            try
            {
                var jsonRequestBody = JsonSerializer.Serialize(accessTokenRequest);
                logger.LogInformation("Getting Account access token: {0}",jsonRequestBody);
                var httpContent = new StringContent(jsonRequestBody, System.Text.Encoding.UTF8, "application/json");

                // Set request uri
                var requestUri = $"{AzureResourceManager}/subscriptions/{SubscriptionId}/resourcegroups/{ResourceGroup}/providers/Microsoft.VideoIndexer/accounts/{ViAccountName}/generateAccessToken?api-version={ApiVersion}";
                var client = new HttpClient(new HttpClientHandler());
                client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", armAccessToken);

                var result = await client.PostAsync(requestUri, httpContent, ct);
                result.EnsureSuccessStatusCode();
                var jsonResponseBody = await result.Content.ReadAsStringAsync(ct);
                logger.LogInformation("Got Account access token: {0},{1}",scope,permission);
                return JsonSerializer.Deserialize<GenerateAccessTokenResponse>(jsonResponseBody)?.AccessToken!;
            }
            catch (Exception ex)
            {
                logger.LogError(ex, "Could not get GetAccountAccessTokenAsync");
                throw;
            }
        }

    }
}

File: VideoIndexerClient/account_token_provider.py
import requests
from azure.identity import DefaultAzureCredential

from VideoIndexerClient.Consts import Consts


def get_arm_access_token(consts:Consts) -> str:
    '''
    Get an access token for the Azure Resource Manager
    Make sure you're logged in with `az` first

    :param consts: Consts object
    :return: Access token for the Azure Resource Manager
    '''
    credential = DefaultAzureCredential()
    scope = f"{consts.AzureResourceManager}/.default" 
    token = credential.get_token(scope)
    return token.token


def get_account_access_token_async(consts, arm_access_token, permission_type='Contributor', scope='Account',
                                   video_id=None):
    '''
    Get an access token for the Video Indexer account
    
    :param consts: Consts object
    :param arm_access_token: Access token for the Azure Resource Manager
    :param permission_type: Permission type for the access token
    :param scope: Scope for the access token
    :param video_id: Video ID for the access token, if scope is Video. Otherwise, not required
    :return: Access token for the Video Indexer account
    '''

    headers = {
        'Authorization': 'Bearer ' + arm_access_token,
        'Content-Type': 'application/json'
    }

    url = f'{consts.AzureResourceManager}/subscriptions/{consts.SubscriptionId}/resourceGroups/{consts.ResourceGroup}' + \
          f'/providers/Microsoft.VideoIndexer/accounts/{consts.AccountName}/generateAccessToken?api-version={consts.ApiVersion}'

    params = {
        'permissionType': permission_type,
        'scope': scope
    }
    
    if video_id is not None:
        params['videoId'] = video_id

    response = requests.post(url, json=params, headers=headers)
    
    # check if the response is valid
    response.raise_for_status()
    
    access_token = response.json().get('accessToken')

    return access_token